python/tchannel/tornado/stream.py

```diff
@@ -0,0 +1,143 @@
+import tornado
+import tornado.gen
+import tornado.ioloop
+from tornado.iostream import PipeIOStream, StreamClosedError
+from ..exceptions import StreamingException
+from ..messages.common import StreamState
+from ..messages import common
+
+try:
+    from tornado.locks import Condition
+except ImportError:  # pragma: no cover
+    from toro import Condition
+
+
+class ArgStream(object):
+
+    def __init__(self):
+        self._state = StreamState.init
+
+    def read(self):
+        """Async read from internal stream buffer
+
+        if it reaches the end of the stream, it will return empty(""), caller
+        will depend on the return value to tell if it reaches the end of the
+        stream.
+
+        It doesn't support seek functionality, which means it will only read
+        the stream data once in single direction.
+
+        NOTE: The data which has been read will be discarded in the stream.
+
+        :return:chunk of bytes read from stream
+        """
+        raise NotImplementedError()
+
+    def write(self, chunk):
+        """Async write to internal stream buffer
+
+        :return chunk of data in the stream
+
+        :raises StreamingException:
+            if stream has been closed, it will raise StreamingException
+        """
+        raise NotImplementedError()
+
+    def close(self):
+        raise NotImplementedError()
+
+
+class InMemArgStream(ArgStream):
```

Comment:
Naming nit: these implementations don't have anything to do with args, they're just reading/writing buffers.